plugin/quick_scope.vim

```diff
@@ -34,437 +32,97 @@ if !exists('g:qs_enable')
-    call add(s:chars_s, char_s)
-  endif
-endfunction
-
-" Set or append to the pattern strings for the highlights.
-function! s:add_to_highlight_patterns(patterns, highlights)
-  let [patt_p, patt_s] = a:patterns
-  let [hi_p, hi_s] = a:highlights
-
-  " If there is a primary highlight for the last word, add it to the primary
-  " highlight pattern.
-  if hi_p > 0
-    let patt_p = printf("%s|%%%sc", patt_p, hi_p)
-  elseif hi_s > 0
-    let patt_s = printf("%s|%%%sc", patt_s, hi_s)
-  endif
-
-  return [patt_p, patt_s]
-endfunction
-
-" Finds which characters to highlight and returns their column positions as a
-" pattern string.
-function! s:get_highlight_patterns(line, start, end, targets)
-  " Keeps track of the number of occurrences for each target
-  let occurrences = {}
-
-  " Patterns to match the characters that will be marked with primary and
-  " secondary highlight groups, respectively
-  let [patt_p, patt_s] = ['', '']
-
-  " Indicates whether this is the first word under the cursor. We don't want
-  " to highlight any characters in it.
-  let is_first_word = 1
-
-  " The positions of the (next) characters that will be given a highlight. A
-  " value of 0 indicates there is no character to highlight.
-  let [hi_p, hi_s] = [0, 0]
-
-  " The (next) characters that will be given a highlight. Used by
-  " save_chars_with_secondary_highlights() to see whether an extra highlight
-  " should be triggered if g:qs_highlight_on_keys is active.
-  let [char_p, char_s] = ['', '']
-
-  " If 1, we're looping forwards from the cursor to the end of the line;
-  " otherwise, we're looping from the cursor to the beginning of the line.
-  let direction = a:start < a:end ? 1 : 0
-
-  let i = a:start
-  while i != a:end
-    let char = a:line[i]
-
-    " Don't consider the character for highlighting, but mark the position
-    " as the start of a new word.
-    "
-    " Check for a <space> as a first condition for optimization.
-    if char == "\<space>" || !has_key(a:targets, char) || empty(char)
-      if !is_first_word
-        let [patt_p, patt_s] = s:add_to_highlight_patterns([patt_p, patt_s], [hi_p, hi_s])
-
-        if exists('g:qs_highlight_on_keys')
-          call s:save_chars_with_secondary_highlights([char_p, char_s])
-        endif
-      endif
-
-      " We've reached a new word, so reset any highlights.
-      let [hi_p, hi_s] = [0, 0]
-      let [char_p, char_s] = ['', '']
-
-      let is_first_word = 0
-    else
-      if has_key(occurrences, char)
-        let occurrences[char] += 1
+  let g:qs_hi_group_cursor = 'QuickScopeCursor'
+
+  if exists('g:qs_first_occurrence_highlight_color')
+    " backwards compatibility mode for old highlight configuration
+    augroup quick_scope_lazy_print
+      if has('vim_starting')
+        " register this as a lazy print error so as not to block Vim starting
+        autocmd CursorHold,CursorHoldI * call quick_scope#lazy_print#err('option g:qs_first_occurrence_highlight_color is deprecated!')
```

Comment:
I think I would prefer a one-time message at Vim's startup rather than a continuous message on `CursorHold`.

Reply:
This does still only print once, it does it in a bit of a convoluted way so as not to block Vim starting up. In the commit message for 9119b8f4ca637d0100f40c0190b5a09b321ca997 I tried to explain it pretty well and also referenced where I learned about it from. Maybe changing to use `echomsg` will remove the need for this trick though, and I would prefer simpler code if we can manage.